components/supervisor/pkg/supervisor/notification.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package supervisor
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	maxPendingNotifications = 100
+)
+
+func NewNotificationService() *NotificationService {
+	return &NotificationService{
+		pendingNotifications: make(map[uint64]*pendingNotification),
+	}
+}
+
+// NotificationService implements the notification service API
+type NotificationService struct {
+	mutex                sync.Mutex
+	subscribers          []api.NotificationService_SubscribeServer
+	nextNotificationId   uint64
+	pendingNotifications map[uint64]*pendingNotification
+}
+
+type pendingNotification struct {
+	message         *api.SubscribeResult
+	responseChannel chan *api.NotifyResponse
+}
+
+// RegisterGRPC registers a gRPC service
+func (srv *NotificationService) RegisterGRPC(s *grpc.Server) {
+	api.RegisterNotificationServiceServer(s, srv)
+}
+
+// RegisterREST registers a REST service
+func (srv *NotificationService) RegisterREST(mux *runtime.ServeMux, grpcEndpoint string) error {
+	return api.RegisterNotificationServiceHandlerFromEndpoint(context.Background(), mux, grpcEndpoint, []grpc.DialOption{grpc.WithInsecure()})
+}
+
+// Sends a notification to the user
+func (srv *NotificationService) Notify(ctx context.Context, req *api.NotifyRequest) (*api.NotifyResponse, error) {
+	if srv.nextNotificationId >= maxPendingNotifications {
+		return nil, status.Error(codes.Internal, "Max number of pending notifications exceeded")
+	}
+	var channel = srv.notifySubscribers(req)
+	if len(req.Actions) == 0 {
+		return &api.NotifyResponse{}, nil
+	}
```

Comment:
You could move this decision to `notifySubscribers` which would always return a channel. If there are no actions, it would simply write `&api.NotifyResponse{}` to the channel and close it immediately.